ebpf_prog/opensnitch.c

```diff
@@ -0,0 +1,446 @@
+	}
+	bpf_probe_read(&udp_key.sport, sizeof(udp_key.sport), &sk->__sk_common.skc_num);
+	bpf_probe_read(&udp_key.saddr, sizeof(udp_key.saddr), &sk->__sk_common.skc_rcv_saddr);
+	
+	u32 zero_key = 0;
+    __builtin_memset(&zero_key, 0, sizeof(zero_key));
+	u64 *counterVal = bpf_map_lookup_elem(&udpcounter, &zero_key);
+	if (counterVal == NULL){return 0;}
+	struct udp_value_t *lookedupValue = bpf_map_lookup_elem(&udpMap, &udp_key);
+	u64 pid = bpf_get_current_pid_tgid() >> 32;
+	if ( lookedupValue == NULL || lookedupValue->pid != pid) {
+		struct udp_value_t udp_value;
+        __builtin_memset(&udp_value, 0, sizeof(udp_value));
+		udp_value.pid = pid;
+		udp_value.uid = bpf_get_current_uid_gid() >> 32;
+		udp_value.counter = *counterVal;
+		bpf_map_update_elem(&udpMap, &udp_key, &udp_value, BPF_ANY);
+		u64 newval = *counterVal + 1;
+		bpf_map_update_elem(&udpcounter, &zero_key, &newval, BPF_ANY);
+	}
+	//else nothing to do
+	return 0;
+
+};
+
+
+SEC("kprobe/udpv6_sendmsg")
+int kprobe__udpv6_sendmsg(struct pt_regs *ctx)
+{	
+	#ifdef OPENSNITCH_x86_32
+		struct sock *sk = (struct sock *)((ctx)->ax);
+		struct msghdr *msg = (struct msghdr *)((ctx)->dx);
+	#else
+		struct sock *sk = (struct sock *)PT_REGS_PARM1(ctx);
+		struct msghdr *msg = (struct msghdr *)PT_REGS_PARM2(ctx);
+	#endif
+
+	u64 msg_name; //a pointer
+    __builtin_memset(&msg_name, 0, sizeof(msg_name));
+	bpf_probe_read(&msg_name, sizeof(msg_name), &msg->msg_name);
+
+	struct udpv6_key_t udpv6_key;
+    __builtin_memset(&udpv6_key, 0, sizeof(udpv6_key));
+	bpf_probe_read(&udpv6_key.dport, sizeof(udpv6_key.dport), &sk->__sk_common.skc_dport);
+	if (udpv6_key.dport != 0){ //likely
+		bpf_probe_read(&udpv6_key.daddr, sizeof(udpv6_key.daddr), &sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
+	}
+	else {
+		struct sockaddr_in6 * sin6 = (struct sockaddr_in6 *)msg_name;
+		bpf_probe_read(&udpv6_key.dport, sizeof(udpv6_key.dport), &sin6->sin6_port);
+		bpf_probe_read(&udpv6_key.daddr, sizeof(udpv6_key.daddr), &sin6->sin6_addr.in6_u.u6_addr32);
+	}
+
+	bpf_probe_read(&udpv6_key.sport, sizeof(udpv6_key.sport), &sk->__sk_common.skc_num);
+	bpf_probe_read(&udpv6_key.saddr, sizeof(udpv6_key.saddr), &sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+
+
+	#ifdef OPENSNITCH_x86_32
+		struct sock_on_x86_32_t sock;
+    	__builtin_memset(&sock, 0, sizeof(sock));
+    	bpf_probe_read(&sock, sizeof(sock), *(&sk));
+		udpv6_key.daddr = sock.daddr;
+		udpv6_key.saddr = sock.saddr;
+	#endif
+
+	u32 zero_key = 0;
+	u64 *counterVal = bpf_map_lookup_elem(&udpv6counter, &zero_key);
+	if (counterVal == NULL){return 0;}
+	struct udpv6_value_t *lookedupValue = bpf_map_lookup_elem(&udpv6Map, &udpv6_key);
+	u64 pid = bpf_get_current_pid_tgid() >> 32;
+	if ( lookedupValue == NULL || lookedupValue->pid != pid) {
+		struct udpv6_value_t udpv6_value;
+        __builtin_memset(&udpv6_value, 0, sizeof(udpv6_value));
+		udpv6_value.pid = pid;
+		udpv6_value.uid = bpf_get_current_uid_gid() >> 32;
```

Comment:
for TCP connections it seems to work fine, but for UDP connections it doesn't obtain the UID. It's either empty or a random int.

Reply:
I println my uid on udp connections while torrenting, I only see uid 1000, nothing is off for me.